### PR TITLE
FE-733: Introducing streaming support

### PR DESCRIPTION
**Pull request by erickpintor**

### CODE REVIEW

File: src/Client.js [hunk 1]
```diff
@@ -16,10 +17,103 @@ var values = require('./values')
...
  */
 
+/**
+ * **WARNING: This is an experimental feature. There are no guarantees to
+ * its API stability and/or service availability. DO NOT USE IT IN
+ * PRODUCTION**.
+ *
+ * Creates a subscription to the result of the given read-only expression. When
+ * executed, the expression must only perform reads and produce a single
+ * streamable type, such as a reference or a version. Expressions that attempt
```
Comment (marrony) [inline src/Client.js:27]: what `version` means in that context? the same as in core?

File: src/stream.js [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+    })
+}
+
+/** Initiates the stream subscription.  */
+StreamClient.prototype.subscribe = function() {
+  var self = this
+  if (self._state === 'idle') {
+    self._state = 'open'
+  } else {
+    throw new Error('The stream alread started at once.')
```
Comment (marrony) [inline src/stream.js:113]: `already`

Comment (faunaee) [reply src/stream.js:113]: "At once" is a synonym for "immediately". Is that what the error message is trying to convey? Or do you mean "The stream has already been started." ?

File: src/stream.js [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+    if (util.isNodeEnv() || fetch.override || !fetch.polyfill) {
+      return null // no override needed
+    }
+
+    fetch = http.resolveFetch(null, false)
+    if (fetch === null) {
+      fetch = function() {
+        return Promise.reject(
+          new errors.StreamsNotSupported(
+            'Could not find a stream compatible fetch function. ' +
```
Comment (faunaee) [inline src/stream.js:73]:
```suggestion
            'Could not find a stream-compatible fetch function. ' +
```

File: src/stream.js [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+              'Please, consider providing a Fetch API compatible function ' +
```
Comment (faunaee) [inline src/stream.js:74]:
```suggestion
              'Please, consider providing a Fetch API-compatible function ' +
```

File: src/stream.js [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+
+/** Subscribe to an event
+ *
+ * @param {string} type The event type.
+ * @param {module:stream~Subscription~eventCalllback} callback
+ *   The event's callback.
+ */
+EventDispatcher.prototype.on = function(type, callback) {
+  if (this._allowedEvents.indexOf(type) === -1) {
+    throw new Error('Unkown event type: ' + type)
```
Comment (faunaee) [inline src/stream.js:265]:
```suggestion
    throw new Error('Unknown event type: ' + type)
```

File: src/stream.js [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+EventDispatcher.prototype.dispatch = function(event) {
+  var listeners = this._listeners[event.event] || []
+  for (var i = 0; i < listeners.length; i++) {
+    listeners[i].call(null, event.data, event)
+  }
+}
+
+/**
+ * Stream's start event. A stream subscription always begins with a start event.
+ * Upcoming events are guaranteed have transaction timestamps equal or grater to
```
Comment (faunaee) [inline src/stream.js:287]:
```suggestion
 * Upcoming events are guaranteed to have transaction timestamps equal to or greater than
```

File: src/stream.js [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+ *   The event type.
+ * @property {number} txnTS
+ *   The event's transaction timestamp.
+ * @property {object} data
+ *   The event's data.
+ */
+
+/**
+ * A snapshot event. A snapshot event is fired once the `document` stream helper
+ * finishes loading the subscribed document's snapshot data load. See {@link
```
Comment (faunaee) [inline src/stream.js:330]:
```suggestion
 * finishes loading the subscribed document's snapshot data. See {@link
```

File: src/stream.js [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+/**
+ * The callback to be executed when an new event occurs.
+ *
+ * @callback module:stream~Subscription~eventCalllback
+ * @param {any} data The event's data field.
+ * @param {object} event The event's entire object.
+ */
+
+/**
+ * A stream subscription which dispatch events received to the registered
```
Comment (faunaee) [inline src/stream.js:379]:
```suggestion
 * A stream subscription which dispatches events received to the registered
```

File: src/stream.js [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+ * Initiates the underlying subscription network calls.
+ * @returns {module:stream~Subscription} This instance.
+ */
+Subscription.prototype.start = function() {
+  this._client.subscribe()
+  return this
+}
+
+/**
+ * Stops the current subscription and close the underlying network connection.
```
Comment (faunaee) [inline src/stream.js:419]:
```suggestion
 * Stops the current subscription and closes the underlying network connection.
```

File: src/Client.js [hunk 10]
```diff
@@ -16,10 +17,103 @@ var values = require('./values')
...
+ * perform writes or produce non-streamable types will result in an error.
```
Comment (faunaee) [inline src/Client.js:28]:
```suggestion
 * to perform writes or produce non-streamable types will result in an error.
```

File: src/stream.js [hunk 11]
```diff
@@ -0,0 +1,488 @@
+'use strict'
+
+/** @module stream */
+
+// NOTE: Although implemented in a separate module, streaming shares internal
+// responsibilities with both Client and HTTP interfaces, such as updating last
+// seen transaction timestamp, and determining if the fetch function set in the
+// client is appropriate for the current runtime. Therefore, this implementation
+// sometimes break encapsulation and call internal getters and methods. As a
```
Comment (faunaee) [inline src/stream.js:9]:
```suggestion
// sometimes breaks encapsulation and calls internal getters and methods. As a
```

File: src/stream.js [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+// general rule: it's okay to call for internal methods. You can interpret this
+// as calling for a package visible method in languages with fine grained
```
Comment (faunaee) [inline src/stream.js:11]:
```suggestion
// as calling for a package visible method in languages with fine-grained
```

**Review by faunaee — COMMENTED** [review]
A question about one of the exception messages, and a few spelling and punctuation fixes.

File: src/Client.js [hunk 13]
```diff
@@ -16,10 +17,103 @@ var values = require('./values')
...
+ *   .on('version', version => processVersion(version))
+ *   .start()
+ * ```
+ *
+ * Be aware that streams are not available in all browsers. If the client can't
+ * initiate a stream, an error event with the {@link
+ * module:errors~StreamsNotSupported} error will be emmited.
+ *
+ * To stop a subscription, call the {@link module:stream~Subscription#close}
+ * method.
```
Comment (ryancharris) [inline src/Client.js:93]: @erickpintor It'd be helpful to have a code sample in this comment block like we do for the `.stream()` function above 👍

File: src/errors.js [hunk 14]
```diff
@@ -255,7 +255,58 @@ function UnavailableError(requestResult) {
...
+}
+
+util.inherits(StreamsNotSupported, StreamError)
+
+/**
+ * An Error thrown by the server when something wrong happened with the
+ * subscribed stream.
+ * @constructor
+ * @param {Object} event The error event.
+ * @property {Object} event The error event.
```
Comment (ryancharris) [inline src/errors.js:297]: @erickpintor Do we need both the `@param` and `@property` notations here?

File: src/stream.js [hunk 15]
```diff
@@ -0,0 +1,488 @@
+'use strict'
+
+/** @module stream */
+
+// NOTE: Although implemented in a separate module, streaming shares internal
+// responsibilities with both Client and HTTP interfaces, such as updating last
+// seen transaction timestamp, and determining if the fetch function set in the
+// client is appropriate for the current runtime. Therefore, this implementation
+// sometimes break encapsulation and call internal getters and methods. As a
+// general rule: it's okay to call for internal methods. You can interpret this
```
Comment (ryancharris) [inline src/stream.js:10]:
```suggestion
// general rule: it's okay to call internal methods. You can interpret this
```

File: src/stream.js [hunk 16]
```diff
@@ -0,0 +1,488 @@
...
+ * to use the native `fetch` API on browsers, while relying on the underlying
+ * `cross-fetch` polyfill from {@link Client} on NodeJS.
+ *
+ * Known limitations:
+ * * NodeJS uses HTTP1;
+ * * Only modern browsers are supported.
+ *
+ * @constructor
+ * @param {Client} client The FaunaDB client.
+ * @param {module:query~ExprArg} expression The fauna expression to subscribe.
```
Comment (ryancharris) [inline src/stream.js:38]:
```suggestion
 * @param {module:query~ExprArg} expression The FQL expression you are subscribing to.
```

File: src/stream.js [hunk 17]
```diff
@@ -0,0 +1,488 @@
...
+ * @param {module:stream~Options} options The stream options.
+ * @param {function} onEvent The stream event's callback.
```
Comment (ryancharris) [inline src/stream.js:40]:
```suggestion
 * @param {function} onEvent The stream event's callback function.
```

File: src/stream.js (hunk 2 above)
Comment (ryancharris) [inline src/stream.js:113]:
```suggestion
    throw new Error('This subscription has already started streaming.')
```

File: src/stream.js [hunk 18]
```diff
@@ -0,0 +1,488 @@
...
+    // close() on a Subscription. There's no need to relay this event back up.
+    if (error.name !== 'AbortError') {
+      self._onEvent({
+        event: 'error',
+        data: error,
+      })
+    }
+  }
+
+  // Minimum browser compatibility based on current code:
```
Comment (ryancharris) [inline src/stream.js:181]: @erickpintor Great documentation here 🚀

File: src/stream.js [hunk 19]
```diff
@@ -0,0 +1,488 @@
...
+  //   Opera for Android     41
+  //   Safari on iOS         10.3
+  //   Samsung Internet      6.0
+  function platformSpecificEventRead(response) {
+    try {
+      if (util.isNodeEnv()) {
+        response.body.on('data', onData).on('error', onError)
+      } else {
+        // ATENTION: The following code is meant to run on browsers and it's not
+        // covered by current test automation. Manual testing on major browsers
```
Comment (ryancharris) [inline src/stream.js:200]:
```suggestion
        // ATENTION: The following code is meant to run in browsers and is not
        // covered by current test automation. Manual testing on major browsers
```

File: src/stream.js [hunk 20]
```diff
@@ -0,0 +1,488 @@
...
+}
+
+/**
+ * Dispatch the given event to the appropriate listeners.
+ *
+ * @param {Object} event The event.
+ */
+EventDispatcher.prototype.dispatch = function(event) {
+  var listeners = this._listeners[event.event] || []
+  for (var i = 0; i < listeners.length; i++) {
```
Comment (ryancharris) [inline src/stream.js:280]: @erickpintor If `listeners` is an empty array, can we do an early return here and avoid the `for` altogether?

**Review by ryancharris — APPROVED** [review]
Looks great! Really nice work on all the extensive documentation and tests 👍

Just left a few questions and suggestions re: wording.

File: src/Client.js (hunk 1 above)
Comment (erickpintor) [reply src/Client.js:27]: Yes. It means that both `Ref(..)` and `Get(Ref(..))` are streamable types.

File: src/stream.js (hunk 2 above)
Comment (erickpintor) [reply src/stream.js:113]: The latter.

**Review by ryancharris — APPROVED** [review]

File: src/stream.js (hunk 18 above)
Comment (marrony) [reply src/stream.js:181]: have you tested all those browsers?

**Review by marrony — APPROVED** [review]

File: src/stream.js (hunk 18 above)
Comment (erickpintor) [reply src/stream.js:181]: Unfortunately, no. I don't have access to all those platforms. This information is derived by the functions called in the code and their minimum browser version documented for them at MDN web docs.

**Pull request merged**